model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java

```diff
@@ -0,0 +1,94 @@
+package org.keycloak.models.map.storage;
+
+import org.keycloak.storage.SearchableModelField;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class QueryParameters<M> {
+
+    protected final Integer offset;
+    protected final Integer limit;
+    protected final List<OrderingTuple<M>> ordering = new LinkedList<>();
+
+    public QueryParameters(Integer offset, Integer limit, List<OrderingTuple<M>> ordering) {
+        this.offset = offset;
+        this.limit = limit;
+        this.ordering.addAll(ordering);
+    }
+
+    public Integer getOffset() {
+        return offset;
+    }
+
+    public Integer getLimit() {
+        return limit;
+    }
+
+    public List<OrderingTuple<M>> getOrdering() {
+        return ordering;
+    }
+
+    enum Order {
+        ASCENDING,
+        DESCENDING
+    }
+
+    public static class OrderingTuple<M> {
+        private final SearchableModelField<M> modelField;
+        private final Order order;
+
+        public OrderingTuple(SearchableModelField<M> modelField, Order order) {
+            this.modelField = modelField;
+            this.order = order;
+        }
+
+        public SearchableModelField<M> getModelField() {
+            return modelField;
+        }
+
+        public Order getOrder() {
+            return order;
+        }
+    }
+
+    public static class Builder<M> {
```

Comment:
Do we want to have `QueryParameters` immutable? I thought it could save us some problems in the future, that is why I created the builder.

Reply:
Not necessarily, we can make it constant

Reply:
What do you mean by constant? Do you mean that we need to initialize all parameters in the constructor. If yes then setting `orderBy` may be complicated. That is why I introduced Builder to make it simple like `.orderBy(1field).orderBy(2field, DESC).orderBy(3field, DESC)` etc. 

We can, of course, do without Builder I just want to understand why we don't want it.

Reply:
My point was slightly elsewhere:

The `Builder` and all of the calls to `store.getQueryParametersBuilder().[...].build()` are a boilerplate code that should not be necessary. Ideally, the whole part would be short, like `withCriteria` or `orderBy` above which makes the code more readable and maintainable. Also, there is nothing storage-specific in the query parameters (it's all about model fields and numbers), so there is no need to have `MapStorage` implement another method. Sidenote - if we find the way to do it reasonably with query parameters, this can lead to refactor of `ModelCriteriaBuilder` use within `MapStorage` as well.

> What do you mean by constant?

Apologies, I should have explained precisely. By a constant I mean `public static final` field. Immutability is not strictly necessary (even though it would be nice).

> I just want to understand why we don't want it.

What is the maximum level of `orderBy` that we really need to support Keycloak? With order, we don't need to be fully generic, it is enough if we support the reasonable usecases.
Builder is excellent pattern for initialization of an immutable object with many fields. Neither pagination nor ordering has that many fields so usage of builder pattern rather obscures the use.